Bugzilla/Auth/Login/APIKey.pm

```diff
@@ -39,36 +38,57 @@ sub app_id {
 # be used to authenticate a Web request.
 sub get_login_info {
   my ($self) = @_;
+  my $cgi    = Bugzilla->cgi;
   my $params = Bugzilla->input_params;
   my ($user_id, $login_cookie);
 
-  my $api_key_text = trim(delete $params->{'Bugzilla_api_key'});
-  if (!i_am_webservice() || !$api_key_text) {
-    return {failure => AUTH_NODATA};
-  }
-
-  my $api_key   = Bugzilla::User::APIKey->new({name => $api_key_text});
-  my $remote_ip = remote_ip();
+  # First check for an API key in the header or passed as query params
+  my $api_key_text = trim($cgi->http('X_BUGZILLA_API_KEY'))
+    || trim($params->{'Bugzilla_api_key'});
 
-  if (!$api_key or $api_key->api_key ne $api_key_text) {
+  # Also allow use of OAuth2 bearer tokens to access the API
+  # substr() removes the "Bearer " portion of the header
+  my $oauth_token = trim($cgi->http('Authorization')) || '';
+  $oauth_token = substr($oauth_token, 7) if $oauth_token;
 
-    # The second part checks the correct capitalization. Silly MySQL
-    ThrowUserError("api_key_not_valid");
-  }
-  elsif ( $api_key->sticky
-    && $api_key->last_used_ip
-    && $api_key->last_used_ip ne $remote_ip)
-  {
-    ThrowUserError("api_key_not_valid");
+  if (!i_am_webservice() || (!$api_key_text && !$oauth_token)) {
+    return {failure => AUTH_NODATA};
   }
-  elsif ($api_key->revoked) {
-    ThrowUserError('api_key_revoked');
+
+  if ($api_key_text) {
+    my $api_key   = Bugzilla::User::APIKey->new({name => $api_key_text});
+    my $remote_ip = remote_ip();
+
+    if (!$api_key or $api_key->api_key ne $api_key_text) {
+
+      # The second part checks the correct capitalization. Silly MySQL
+      ThrowUserError("api_key_not_valid");
+    }
+    elsif ($api_key->sticky
+      && $api_key->last_used_ip
+      && $api_key->last_used_ip ne $remote_ip)
+    {
+      ThrowUserError("api_key_not_valid");
+    }
+    elsif ($api_key->revoked) {
+      ThrowUserError('api_key_revoked');
+    }
+
+    $api_key->update_last_used($remote_ip);
+    $self->set_app_id($api_key->app_id);
+
+    return {user_id => $api_key->user_id};
   }
 
-  $api_key->update_last_used($remote_ip);
-  $self->set_app_id($api_key->app_id);
+  if ($oauth_token) {
```

Comment:
At first glance there doesn't appear to be any validation of `$oauth_token`.

Please add a comment pointing to `Bugzilla::App::Plugin::OAuth2` where the validation occurs.